gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java

```diff
@@ -146,14 +134,40 @@ public ApiKeyEntity renew(String subscription, String customApiKey) {
         }
     }
 
-    /**
-     * Generate an {@link ApiKey} from a subscription
-     *
-     * @param subscription
-     * @return An Api Key
-     */
-    private ApiKey generateForSubscription(String subscription) {
-        return generateForSubscription(subscription, null);
+    private ApiKeyEntity findOrGenerate(ApplicationEntity application, SubscriptionEntity subscription, String customApiKey) {
+        return findByApplication(application.getId())
+            .stream()
+            .findFirst()
+            .map(
+                apiKey -> {
+                    apiKey.addSubscription(subscription);
```

Comment:
I think we should add an audit log that indicates that a subscription has been added to a shared api key.
@mouligno WDYT ? Maybe we could plan a dedicated ticket for audit logs regarding shared api keys ?

Reply:
dedicated ticket has been created